### PR TITLE
[stable/elasticsearch] Elasticsearch v7.1.1

### LINKED ISSUE #14376: [stable/elasticsearch] support 6.8 and 7.1 es versions
Add support for 6.8 and 7.1 versions of Elasticsearch.
Latest releases support [many interesting features](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/release-highlights-6.8.0.html).
Current version of the chart does not seem to support versions newer than 6.7.

**Describe the solution you'd like**
Support newer versions, including added features (most notably: RBAC)

### PR DESCRIPTION
**Pull request by tetianakravchenko**
#### What this PR does / why we need it:

- switch to es version 7.1.1
- adjust accordingly `elasticsearch.yaml` ( add `cluster.initial_master_nodes`)
- add `log4j2.properties` for ESv7

- I've also added `(hasPrefix "6." .Values.appVersion)` for mounting `log4j2.properties` and setting `NODE_INGEST`, seems like that was missed.

#### Which issue this PR fixes
  - fixes https://github.com/helm/charts/issues/14376

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

File: stable/elasticsearch/Chart.yaml [hunk 1]
```diff
@@ -1,8 +1,8 @@
 apiVersion: v1
 name: elasticsearch
 home: https://www.elastic.co/products/elasticsearch
-version: 1.28.0
-appVersion: 6.7.0
+version: 1.29.0
```
Comment (t-d-d) [inline stable/elasticsearch/Chart.yaml:4]: Note that you can't do a rolling update from 6.7 to 7.* .  So should this be a major version bump?  I'm not sure.  But at least there should be some documentation in the readme.  https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-upgrade.html

Comment (tetianakravchenko) [reply stable/elasticsearch/Chart.yaml:4]: @t-d-d thank you for pointing it out! I've added readme and bumped version to `6.8`, as I mentioned it in the readme as a version supported to perform update (6.8 to 7.1) https://github.com/helm/charts/pull/14535/commits/713768d2d470e89c492e75c2dc7dd6b7ee048c70

**Comment by Towmeykaw** [comment]
We talked about this change on a meeting regarding the depreciation of this chart in favor of the official elasticsearch chart. The official chart already support v7 and the question is if a restart is required then it might be a good time to switch to that chart. Is something stopping people from switching which may make this upgrade needed?

**Comment by anas-aso** [comment]
@Towmeykaw from the official chart : https://github.com/elastic/helm-charts/tree/master/elasticsearch
> This functionality is in beta status and may be changed or removed completely in a future release. 

I believe this is enough for a reason to not switch to the official chart (yet). In addition, the official chart doesn't support preStop/preStart hook which we need in our case.

**Comment by Crazybus** [comment]
> This functionality is in beta status and may be changed or removed completely in a future release.

> I believe this is enough for a reason to not switch to the official chart (yet).

Thanks for calling this out. I bumped the status from alpha to beta before I had the official messaging ready (to have it done on time for Kubecon). I have just submitted a PR (https://github.com/elastic/helm-charts/pull/158) to update this. This is the official messaging that I need to put in, however I'm fully committed to following semver and not making breaking changes between major releases. Only one breaking change was made when going from 6.x to 7.x and this was a simple values.yaml change to remove the default storage class name (https://github.com/elastic/helm-charts/releases/tag/7.0.0-alpha1). The only difference between this being in beta and stable now is the support you get if you are an Elastic customer. 

> In addition, the official chart doesn't support preStop/preStart hook which we need in our case.

The Elastic helm chart uses a [different approach](https://github.com/elastic/helm-charts/blob/ecdfd66956ffd4f8a27237a51b9c8644e30c7f8a/elasticsearch/templates/statefulset.yaml#L151-L163) which I have found to be much more reliable. Instead of attempting to fully drain a node when stopping and starting, it instead waits for the cluster to be fully green during initial startup. The big advantage of this is that you can easily rollout changes without needing to sync all data to a new node and back again. This can be combined with setting `"index.unassigned.node_left.delayed_timeout": "5m"` to make sure that shards are never re-synced unnecessarily during rolling restarts. 

Another issue with using lifecycle hooks for this is with how the [execution is handled in Kubernetes](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-handler-execution) which makes them not a very reliable thing to use for something as important as this. If a single call fails then it means that a node might not be added back into the cluster. 

The current grep is also going to cause issues for clusters with more than 10 nodes.

https://github.com/helm/charts/blob/3b8631a4dbdbc7585ff3e026bea56ef9e3785f76/stable/elasticsearch/templates/configmap.yaml#L117

If you have 10 nodes then a grep for `master-1` is also going to match for `master-11` which could lead to some unexpected results.

**Comment by t-d-d** [comment]
@Towmeykaw But it seems you can update from 6.8 to 7.
If this chart is to be deprecated my vote would be to make that clear at the top of the ReadMe and allow only none-breaking PRs for a time.  The intend would be to give users of this chart some notice and a chance to switch over to the Elastic chart.

**Comment by anas-aso** [comment]
@Crazybus Thanks for the details clarification. I agree with @t-d-d that it should be clear that this chart won't support any newer ES version and that it's in maintenance mode only.

**Comment by jethr0null** [comment]
@t-d-d @anas-aso We were actually chatting about this just last week. Here is the [PR](https://github.com/helm/charts/pull/14763/files) with the initial messaging letting folks know about the intent to move over to the Elastic maintained Charts. The added text also includes a link to the migration guide which outlines the process for moving from the community chart to the Elastic chart.

**Comment by tetianakravchenko** [comment]
@Crazybus Thank you for the clarification!

I will close this PR.

**Comment by tothdavid** [comment]
@Towmeykaw I believe there are some unaddressed issues about migrating to the official elasticsearch helm chart: These may not seem very important for some but actually this helm chart (stable/elasticsearch) follows a very different approach than the one in the elastic repo. That is the official elastic helm chart sets up only one role of a cluster, which means that you need to create a separate helm chart installation for master, client, data, or whatever roles you want in your cluster. Also you need to make sure that the different parts of your cluster will be able to see each other as this can not be provided by the official helm chart because by its nature one helm chart instance is only aware of one role of the cluster.
Also there is another issue: this stable/elasticsearch chart is used by the stable/elastic-stack chart which will setup an entire Elastic Stack with elasticsearch, logstash, kibana, and whatever other components you may want. So if you make this chart deprecated and will not support anything newer than 6.7 that means that you basically deprecate the elastic-stack chart as well (and actually there is no notice of this issue there at all).
So to make this work like before the elastic-stack chart also needs to be updated to utilize the official elasticsearch chart either in a way that it would use it in 3 instances (master, client, data) or even there could be an intermediate chart created (e.g. elasticsearch-cluster) that would take care of this aspect and the elastic-stack could just depend on this intermediate one. Also there is no official kibana, logstash or other similar charts in that repo so we still need to use the ones here.
On the other hand it would be much-much easier to just merge this pull request which could provide support for newer elasticsearch versions for us who are fine with this chart and could continue using elastic-stack chart as well for the time while there is nothing similar provided by Elastic or while elastic-stack chart would not support the new chart.

**Comment by hectorj2f** [comment]
> I agree with @t-d-d that it should be clear that this chart won't support any newer ES versions and that it's in maintenance mode only.

I agree with that too. 👍

**Pull request closed without merging**